packages/vite/src/node/cleanString.ts

```diff
@@ -0,0 +1,77 @@
+// bank on the non-overlapping nature of regex matches and combine all filters into one giant regex
+// `([^`\$\{\}]|\$\{(`|\g<1>)*\})*` can match nested string template
+// but js not support match expression(\g<0>). so clean string template(`...`) in other ways.
+const cleanerRE = /"[^"]*"|'[^']*'|\/\*(.|[\r\n])*?\*\/|\/\/.*/g
+
+const blankReplacer = (s: string) => ' '.repeat(s.length)
+const stringBlankReplacer = (s: string) =>
+  `${s[0]}${'\0'.repeat(s.length - 2)}${s[0]}`
+
+export interface CleanString {
+  clean: string
+  raw: string
+}
+
+function isCleanString(obj: any): obj is CleanString {
+  return obj.raw && obj.clean
+}
+
+export function emptyCommentsString(raw: string): CleanString {
+  const res: CleanString = {
+    raw: raw,
+    clean: raw.replace(cleanerRE, (s: string) =>
+      s[0] === '/' ? blankReplacer(s) : s
+    )
+  }
+  return res
+}
+
+export function emptyString(raw: string | CleanString): CleanString {
+  const res: CleanString = { raw: '', clean: '' }
+  if (isCleanString(raw)) {
+    res.raw = raw.raw
+    res.clean = raw.clean
+  } else {
+    res.raw = raw
+    res.clean = raw
+  }
+  res.clean = res.clean.replace(cleanerRE, (s: string) =>
+    s[0] === '/' ? blankReplacer(s) : stringBlankReplacer(s)
+  )
+  // TODO replace string template
+  return res
+}
+
+export function findEmptyStringRawIndex(
+  raw: CleanString,
+  emptyFlag: string,
+  start: number
+): [number, number] {
+  const flagIndex = raw.clean.indexOf(emptyFlag, start)
+  const flagEndIndex = flagIndex + emptyFlag.length
+  return [flagIndex, flagEndIndex]
+}
+
+export async function walkCleanString(
+  re: RegExp,
+  raw: string,
+  callback: (match: RegExpExecArray, cleanString: CleanString) => Promise<void>
+): Promise<void> {
+  const cleanString = emptyString(raw)
+  let match: RegExpExecArray | null
+  while ((match = re.exec(cleanString.clean))) {
+    await callback(match, cleanString)
+  }
+}
+
+export function walkCleanStringSync(
+  re: RegExp,
+  raw: string,
+  callback: (match: RegExpExecArray, cleanString: CleanString) => void
+): void {
+  const cleanString = emptyString(raw)
+  let match: RegExpExecArray | null
+  while ((match = re.exec(cleanString.clean))) {
+    callback(match, cleanString)
+  }
+}
```

Comment:
@poyoho I'm checking the code after applying these abstractions, but I'm not sure if we are getting much at the end. I think we could have:
```ts
export function emptyCommentsString(raw: string): string {
  return raw.replace(cleanerRE, (s: string) =>
    s[0] === '/' ? blankReplacer(s) : s
  )
}
```
And:
```ts
export function emptyString(raw: string): string {
  return raw.replace(cleanerRE, (s: string) =>
    s[0] === '/' ? blankReplacer(s) : stringBlankReplacer(s)
  )
}
```
We have the `raw` string where we call these functions, so I think the `{raw,clean}` abstraction isn't needed.
I think `findEmptyStringRawIndex` is useful:
```ts
export function findEmptyStringRawIndex(
  clean: string,
  emptyFlag: string,
  start: number
): [number, number] {
  const flagIndex = clean.indexOf(emptyFlag, start)
  const flagEndIndex = flagIndex + emptyFlag.length
  return [flagIndex, flagEndIndex]
}
```
And then we can avoid `walkCleanString` and `walkCleanStringSync`, since they will become only a loop with a callback, and it that case better to have the loop so it is more explicit.
Maybe there are other abstractions to simplify the code, but I think it is better that we keep things simpler and try to merge this PR that has a lot of value in fixing the current bugs and adding proper tests.
What do you think?

Reply:
very useful advice. Because js not support match expression `(\g<0>)` so regexp can't clean string template(\`...\`)

Reply:
Would you explain that a bit more?

What do you think about removing the `{ raw, clean }` abstraction and directly return the clean string from `emptyCommentsString` and `emptyString` (https://github.com/vitejs/vite/pull/7650#discussion_r846674812)?

Reply:
good! Returning a raw may result in repeated creation of a new string, and `raw` is originally provided by the caller and should not require additional returns.